### PR TITLE
Add all source files for the customer client

### PR DESCRIPTION
**Pull request by s-and-witch**

### CODE REVIEW

File: src/Customer/Types/TrackAnonymusEvent.hs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+data TrackAnonymusEventBody = MkTrackAnonymusEvent
+  { taeName        :: Text
+  , taeAnonymousId :: Maybe Text
+  , taeId          :: Maybe Text
+  , taeTimestamp   :: Maybe Int
+  , taeData        :: Maybe Object
+  } deriving stock (Generic)
+
+instance ToJSON TrackAnonymusEventBody where
```
Comment (rvem) [inline src/Customer/Types/TrackAnonymusEvent.hs]: Isn't it the same as `deriveToJSON` from `Data.Aeson.TH`?

TH splice would be a bit shorter and less boilerplaty, what do you think?

File: package.yaml [hunk 2]
```diff
@@ -0,0 +1,120 @@
+# SPDX-FileCopyrightText: 2020 Serokell <https://serokell.io>
+#
+# SPDX-License-Identifier: CC0-1.0
+
+
+name: customer-client
```
Comment (rvem) [inline package.yaml]: I think `customer-client` name might be a bit confusing. One can only find out that this library adds binding for customer.io by reading package meta. 
For example, library with python bindings is named [`customerio`](https://github.com/customerio/customerio-python), I propose renaming it to something like `customerio-client`

File: .github/pull_request_template.md [hunk 3]
```diff
@@ -47,25 +36,3 @@ you must be able to justify that.
...
-[//]: # (Mostly for public repositories)
-[//]: # (Recording changes is optional, depends on repository, useful for some libs)
-- Public contracts
-  - [ ] Any modifications of public contracts comply with the [Evolution
-  of Public Contracts](https://www.notion.so/serokell/Evolution-of-Public-Contracts-2a3bf7971abe4806a24f63c84e7076c5) policy.
-  - [ ] I added an entry to the [changelog](../tree/master/CHANGES.md) if my changes are visible to the users
-        and
-  - [ ] provided a migration guide for breaking changes if possible
-
-#### Stylistic guide (mandatory)
```
Comment (rvem) [inline .github/pull_request_template.md:66]: Hmm, why was this removed?

File: .github/pull_request_template.md [hunk 4]
```diff
@@ -47,25 +36,3 @@ you must be able to justify that.
   - [ ] If I added new functionality, I added tests covering it.
   - [ ] If I fixed a bug, I added a regression test to prevent the bug from
         silently reappearing again.
-
-[//]: # (Add more docs here if you have them in the repository)
```
Comment (rvem) [inline .github/pull_request_template.md:51]: And this

File: flake.nix [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+        haskellPackages = pkgs.haskellPackages;
+
+        jailbreakUnbreak = pkg:
+          pkgs.haskell.lib.doJailbreak (pkg.overrideAttrs (_: { meta = { }; }));
+
+        packageName = "customer-client";
+      in {
+        packages.${packageName} =
+          haskellPackages.callCabal2nix packageName self rec {
```
Comment (rvem) [inline flake.nix]: Is there a specific reason `cabal2nix` was used over `haskell.nix`?

File: src/Customer/Types/TrackCustomerEvent.hs [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+import Data.Aeson (Object, ToJSON(toJSON), genericToJSON)
+import Data.Text (Text)
+import GHC.Generics (Generic)
+
+data TrackCustomerEventBody = MkTrackCustomerEvent
+  { tceName      :: Text
+  , tceId        :: Maybe Text
+  , tceTimestamp :: Maybe Int
+  , tceData      :: Maybe Object
+  }deriving stock (Generic)
```
Comment (rvem) [inline src/Customer/Types/TrackCustomerEvent.hs]:
```suggestion
  } deriving stock (Generic)
```

File: src/Customer/Aeson.hs [hunk 7]
```diff
@@ -0,0 +1,15 @@
+-- SPDX-FileCopyrightText: 2022 Serokell <https://serokell.io/>
+--
+-- SPDX-License-Identifier: Apache-2.0
+
+module Customer.Aeson (defaultAesonOptions) where
+
+import Data.Aeson (Options(fieldLabelModifier), camelTo2, defaultOptions)
+
+defaultAesonOptions :: String -> Options
+defaultAesonOptions prefix = defaultOptions
```
Comment (rvem) [inline src/Customer/Aeson.hs]: Just a nitpick, but I believe the thing you want could be expressed shorter: `aesonPrefix snakeCase`?:)

File: src/Customer/API.hs [hunk 8]
```diff
@@ -0,0 +1,44 @@
+-- SPDX-FileCopyrightText: 2022 Serokell <https://serokell.io/>
+--
+-- SPDX-License-Identifier: Apache-2.0
+
+module Customer.API (api) where
```
Comment (rvem) [inline src/Customer/API.hs]: Since this only provides `/api/v1/customer/{identifier}/events`, I propose renaming this module to `Customer.Events.API`. Also, are there any plans to support more API endpoints?

File: src/Customer/Types/TrackCustomerEvent.hs [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+import Customer.Aeson (defaultAesonOptions)
+import Data.Aeson (Object, ToJSON(toJSON), genericToJSON)
+import Data.Text (Text)
+import GHC.Generics (Generic)
+
+data TrackCustomerEventBody = MkTrackCustomerEvent
+  { tceName      :: Text
+  , tceId        :: Maybe Text
+  , tceTimestamp :: Maybe Int
+  , tceData      :: Maybe Object
```
Comment (rvem) [inline src/Customer/Types/TrackCustomerEvent.hs]: AFAICS, this API endpoint allows any additional payload in JSON format. How hard will it be to make it polymorphic and allow providing an object of type `a`, that has `ToJSON` instance?

File: src/Customer/Aeson.hs (hunk 7 above)
Comment (rvem) [reply src/Customer/Aeson.hs]: And then an instance could be derived via `deriveJSON  defaultAesonOptions ''<type>`

File: src/Customer/Types/ReportPushMetrics.hs [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+  ) where
+
+import Customer.Aeson (defaultAesonOptions)
+import Data.Aeson (ToJSON(toJSON), genericToJSON)
+import Data.Text (Text)
+import GHC.Generics (Generic)
+
+data ReportPushMetricsBody = MkReportPushMetrics
+  { rpmDeliveryId :: Maybe Text
+  , rpmEvent      :: Maybe Text
```
Comment (rvem) [inline src/Customer/Types/ReportPushMetrics.hs]: This is a enum in API description, perhaps, it should be represented as ADT instead of `Text`

**Review by rvem — COMMENTED** [review]
Have you tested that API bindings actually work as expected?

File: flake.nix (hunk 5 above)
Comment (s-and-witch) [reply flake.nix]: > Is there a specific reason cabal2nix was used over haskell.nix?

I had already working setup with cabal2nix and it better matches nixos caches, so it's easier to use as a dependency.

File: src/Customer/API.hs (hunk 8 above)
Comment (s-and-witch) [reply src/Customer/API.hs]: >  Also, are there any plans to support more API endpoints?

I don't know.

File: .github/pull_request_template.md (hunk 3 above)
Comment (s-and-witch) [reply .github/pull_request_template.md:66]: I can restore it, but I'm not sure that it would be very helpful

**Comment by s-and-witch** [comment]
> Have you tested that API bindings actually work as expected?

How can I do that?

File: package.yaml (hunk 2 above)
Comment (s-and-witch) [reply package.yaml]: Should we also change repository link?

File: .github/pull_request_template.md (hunk 4 above)
Comment (s-and-witch) [reply .github/pull_request_template.md:51]: We have no docs in the repository

File: package.yaml (hunk 2 above)
Comment (rvem) [reply package.yaml]: Yes

File: .github/pull_request_template.md (hunk 4 above)
Comment (rvem) [reply .github/pull_request_template.md:51]: But we should probably have haddock :thinking:

File: .github/pull_request_template.md (hunk 3 above)
Comment (rvem) [reply .github/pull_request_template.md:66]: These are development standards that are used within our company

File: .github/pull_request_template.md (hunk 4 above)
Comment (rvem) [reply .github/pull_request_template.md:51]: Also, we should have some README

File: src/Customer.hs [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+import Customer.Events.Types.ReportPushMetrics (ReportPushMetricsBody(..), defaultReportPushMetrics)
+import Customer.Events.Types.TrackAnonymusEvent (TrackAnonymusEventBody(..), defaultTrackAnonymusEvent)
+import Customer.Events.Types.TrackCustomerEvent (TrackCustomerEventBody(..), defaultTrackCustomerEvent)
+import Data.Text (Text)
+import qualified Network.HTTP.Client as HTTP
+import Servant.API
+import Servant.Client
+
+host :: BaseUrl
+host = BaseUrl Https "track.customer.io" 443 ""
```
Comment (rvem) [inline src/Customer.hs:25]: Let's not hardcode URLs, this might make the life of users much harder. One day they'll change the URL and thus it'll be necessary to update the library, which is quite inconvenient

File: src/Customer/Events/Types/ReportPushMetrics.hs [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+instance ToJSON EventType where
+  toJSON = \case
+    Opened -> "opened"
+    Converted -> "converted"
+    Delivered -> "delivered"
+
+data ReportPushMetricsBody = MkReportPushMetrics
+  { rpmDeliveryId :: Maybe Text
+  , rpmEvent      :: Maybe Text
```
Comment (rvem) [inline src/Customer/Events/Types/ReportPushMetrics.hs]: `EventType` was introduced, but seems like it wasn't used

File: src/Customer/Events/Types/ReportPushMetrics.hs [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+
+data EventType = Opened | Converted | Delivered
+
+instance ToJSON EventType where
+  toJSON = \case
+    Opened -> "opened"
+    Converted -> "converted"
+    Delivered -> "delivered"
+
+data ReportPushMetricsBody = MkReportPushMetrics
```
Comment (rvem) [inline src/Customer/Events/Types/ReportPushMetrics.hs]: Documentation at https://customer.io/docs/api/#operation/trackAnonymous is somewhat confusing since they don't explicitly define whether the payload parts of this API endpoint are optional, but from the general semantics is seems like they shouldn't be optional. Though, I'm not that familiar with customer.io API, so please double-check whether they should be optional

File: src/Customer/Events/Types/TrackCustomerEvent.hs [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+import Customer.Aeson (defaultAesonOptions)
+import Data.Aeson (Object)
+import Data.Aeson.TH (deriveToJSON)
+import Data.Text (Text)
+
+data TrackCustomerEventBody = MkTrackCustomerEvent
+  { tceName      :: Text
+  , tceId        :: Maybe Text
+  , tceTimestamp :: Maybe Int
+  , tceData      :: Maybe Object
```
Comment (rvem) [inline src/Customer/Events/Types/TrackCustomerEvent.hs]: Oh, actually, looks like it expects not just an `Object`, but something with a specific structure:
image
This one is from https://customer.io/docs/api/#operation/trackAnonymous

**Comment by rvem** [comment]
> How can I do that?

https://docs.servant.dev/en/stable/cookbook/testing/Testing.html

But first we should probably somehow check that bindings we provide are actually creating valid requests, but I'm not sure if there is a playground for customer.io

File: src/Customer.hs [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+
+trackAnonymusEventC' :: Env -> TrackAnonymusEventBody -> IO (Either ClientError ())
+trackAnonymusEventC' MkEnv{..} body = do
+  runClientM (trackAnonymusEventC authtoken body) clientEnv
+
+reportPushMetricsC' :: Env -> ReportPushMetricsBody -> IO (Either ClientError ())
+reportPushMetricsC' MkEnv{..} body = do
+  runClientM (reportPushMetricsC body) clientEnv
+
+trackCustomerEvent :: Env -> Text -> Text -> IO (Either ClientError ())
```
Comment (rvem) [inline src/Customer.hs]: Can you justify why helpers with `default*Event` bodies can be useful? I can think of a use-case for them when using this library as an SDK

File: src/Customer/API.hs (hunk 8 above)
Comment (rvem) [reply src/Customer/API.hs]: Oh, actually looks like it should be `Customer.Track.Events.API` because there are Track API and App API

File: src/Customer.hs (hunk 15 above)
Comment (s-and-witch) [reply src/Customer.hs]: I'm not sure, that they are useful, but `default*Event` are useful because user can override part of fields and don't touch other. I can remove "the most default" helpers, but in that case I need to know that they aren't useful. Basically, I just applied structure from `mixpanel-client`, that Domen mentioned at the start.

File: src/Customer/API.hs (hunk 8 above)
Comment (s-and-witch) [reply src/Customer/API.hs]: BTW, Domen have no interest in App API

File: src/Customer/Types/TrackCustomerEvent.hs (hunk 9 above)
Comment (s-and-witch) [reply src/Customer/Types/TrackCustomerEvent.hs]: Seems like it is not possible due to necessity to provide ToJSON instance to servant's `client` function. I can make it existential, but there would be problems with making right ToJSON instance.

File: src/Customer/API.hs (hunk 8 above)
Comment (rvem) [reply src/Customer/API.hs]: Yes, but some other people may potentially come across this repository and use it, and it might me confusing for them

File: src/Customer/Track/Events/Types/TrackAnonymusEvent.hs [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+  ( module Customer.Track.Events.Types.TrackAnonymusEvent
+  ) where
+
+import Customer.Aeson (defaultAesonOptions, mkObject, mkPair)
+import Data.Aeson
+import Data.Aeson.TH (deriveToJSON)
+import Data.Text (Text)
+
+data TrackAnonymusEventBody
+  = MkStandardAnonymusEvent
```
Comment (rvem) [inline src/Customer/Track/Events/Types/TrackAnonymusEvent.hs]: Let's not use record syntax with sum types, pls, it's wildly unsafe

File: src/Customer.hs (hunk 15 above)
Comment (rvem) [reply src/Customer.hs]: The structure is good, but I can't think of a use-case for `default*Event` helpers

File: README.md [hunk 17]
```diff
@@ -1,77 +1,24 @@
...
    -->
 
-**Please read this file as a raw textual file, it has comments for people
-who create a new repo from this template.**
+[![License: Apache 2.0](https://img.shields.io/badge/License-Apache%202.0-brightgreen.svg)](https://opensource.org/licenses/Apache-2.0)
 
-This repository is a template that we use at Serokell to create new repositories.
-See [GitHub documentation](https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/creating-a-repository-from-a-template)
-for more details about this feature.
+# Customer.io client for haskell
```
Comment (rvem) [inline README.md:9]: Let's have a bit more meaningful section that will describe what kind of client this repository actually provides. We should probably mention that it is `servant-client` based and currently supports only some subset of `Track` customer.io API

File: .github/workflows/ci.yaml [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+      matrix:
+        cabal:
+          - '3.4'
+        ghc:
+          - '9.2.5'
+          - '9.0.2'
+          - '8.10.7'
+          - '8.8.4'
+  validate:
+    runs-on: ubuntu-latest
```
Comment (rvem) [inline .github/workflows/ci.yaml]: you can use our GA runners with `self-hosted` tag that already has nix installed

File: .github/workflows/ci.yaml [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+    steps:
+      - uses: "actions/checkout@v3"
+      - id: setup-haskell-cabal
+        uses: "haskell/actions/setup@v2"
+        with:
+          cabal-version: "${{ matrix.cabal }}"
+          enable-stack: false
+          ghc-version: "${{ matrix.ghc }}"
+      - uses: cachix/install-nix-action@v18
```
Comment (rvem) [inline .github/workflows/ci.yaml]: Is it needed now? I don't see any nix usage within `build`

File: flake.nix (hunk 5 above)
Comment (rvem) [reply flake.nix]: Btw, would be nice to check that this setup works in CI to avoid breaking it accidentally in future.
Also, AFAIR, one downside of using `cabal2nix` is that it will use dependencies from `haskellPackages` from `nixpkgs`, which may be outdated and/or broken, while `haskell.nix` will use deps from hackage directly

File: README.md [hunk 20]
```diff
@@ -1,77 +1,24 @@
 <!--
    - SPDX-FileCopyrightText: 2019-2021 Serokell <https://serokell.io>
```
Comment (rvem) [inline README.md]:
```suggestion
   - SPDX-FileCopyrightText: 2022 Serokell <https://serokell.io>
```
And presumably `2022-2023`

File: .github/workflows/ci.yaml (hunk 18 above)
Comment (s-and-witch) [reply .github/workflows/ci.yaml]: How can I use it? 
> error: file 'nixpkgs' was not found in the Nix search path (add it using $NIX_PATH or -I)

File: README.md [hunk 21]
```diff
@@ -1,77 +1,33 @@
...
-## For Contributors [↑](#-patak)
+## For Contributors
 
 Please see [CONTRIBUTING.md](CONTRIBUTING.md) for more information.
 
-## About Serokell [↑](#-patak)
+## About Serokell
 
-Patak is maintained and funded with ❤️ by [Serokell](https://serokell.io/).
+Customer-client is maintained and funded with ❤️ by [Serokell](https://serokell.io/).
```
Comment (rvem) [inline README.md]:
```suggestion
CustomerIO-client is maintained and funded with ❤️ by [Serokell](https://serokell.io/).
```

File: .github/workflows/ci.yaml (hunk 18 above)
Comment (rvem) [reply .github/workflows/ci.yaml]: You can either add reuse to the `devShell` and do something like `nix develop .#ci -c reuse lint` or extract reuse into a flake check, we have the corresponding [helper](https://github.com/serokell/serokell.nix/blob/master/overlay/default.nix#L41) in serokell.nix, usage example can be found [here](https://github.com/serokell/templates/blob/master/haskell.nix/flake.nix#L103)

Comment (rvem) [reply .github/workflows/ci.yaml]: > file 'nixpkgs' was not found in the Nix search path 

This is done in order to avoid unexpected breaking changes in `<nixpkgs>`

File: src/Customer/Events/Types/ReportPushMetrics.hs (hunk 13 above)
Comment (rvem) [reply src/Customer/Events/Types/ReportPushMetrics.hs]: I'm looking at the official ruby SDK and seems like these fields indeed cannot be `Nothing`, see https://github.com/customerio/customerio-ruby/blob/main/lib/customerio/client.rb#L91

**Review by rvem — APPROVED** [review]
Okay, I think this PR received a sufficient amount of efforts:)

Please clean up and split changes in at least three commits:
1) That adds/updates files that don't directly relate to the Haskell project
2) That adds Haskell project
3) nix/CI setup

Though, let's not spend too much time on it, if it turns out to be too difficult

File: src/Customer/Track/Events/Types/TrackCustomerEvent.hs [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+import Data.Text (Text)
+
+data TrackCustomerEventBody = MkTrackCustomerEvent
+  { tceName      :: Text
+  , tceId        :: Maybe Text
+  , tceTimestamp :: Maybe Int
+  , tceData      :: Maybe CustomerEventData
+  }
+
+data CustomerEventData = MkCustomerEventData
```
Comment (sandydoo) [inline src/Customer/Track/Events/Types/TrackCustomerEvent.hs:21]: The [API docs for tracking](https://customer.io/docs/api/track/?region=eu#operation/track) are misleading. What they're actually documenting is an arbitrary JSON object, not a structured payload with those attributes.

`data` can contain arbitrary attributes, in addition to a few reserved ones for sending emails: `recipient`, `from_address`, and `reply_to`. The reserved attributes are optional.

File: src/Customer/Track/Events/Types/TrackAnonymusEvent.hs [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+module Customer.Track.Events.Types.TrackAnonymusEvent
+  ( module Customer.Track.Events.Types.TrackAnonymusEvent
+  ) where
+
+import Customer.Aeson (defaultAesonOptions, mkObject, mkPair)
+import Data.Aeson
+import Data.Aeson.TH (deriveToJSON)
+import Data.Text (Text)
+
+data TrackAnonymusEventBody
```
Comment (sandydoo) [inline src/Customer/Track/Events/Types/TrackAnonymusEvent.hs]: `Anonymus` -> `Anonymous`

**Comment by sandydoo** [comment]
Are you interested in merging this? There's a v2 API out now that we might add support for.

File: src/Customer/Track/Events/Types/TrackCustomerEvent.hs (hunk 22 above)
Comment (s-and-witch) [reply src/Customer/Track/Events/Types/TrackCustomerEvent.hs:21]: We capture this behavior in the ToJSON instance. All the additional fields are merging with description/type/properties into one JSON object.

File: src/Customer/Track/Events/Types/TrackAnonymusEvent.hs (hunk 23 above)
Comment (s-and-witch) [reply src/Customer/Track/Events/Types/TrackAnonymusEvent.hs]: Fixed

**Comment by s-and-witch** [comment]
> Are you interested in merging this? There's a v2 API out now that we might add support for.

Currently, we have no capabilities to update these bindings. I can merge this PR into master, if you want.

File: src/Customer/Track/Events/Types/TrackCustomerEvent.hs (hunk 22 above)
Comment (sandydoo) [reply src/Customer/Track/Events/Types/TrackCustomerEvent.hs:21]: Sure, but `description`, `type`, and `properties` are not part of the API. They're *not* reserved fields. It's a mistake in the documentation on customerio.

Comment (s-and-witch) [reply src/Customer/Track/Events/Types/TrackCustomerEvent.hs:21]: Oh, now I understand you

**Comment by sandydoo** [comment]
> Currently, we have no capabilities to update these bindings. I can merge this PR into master, if you want.

A merge would be great!
